### PR TITLE
test: Code Coverage: Fy-Select-Project Modal Component

### PR DESCRIPTION
**Pull request by jayfyle**
### Clickup: https://app.clickup.com/t/85zrxnag7

# Coverage at 98%: 

fy-select-project-modal-98

### CODE REVIEW

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+    recentLocalStorageItemsService = TestBed.inject(
+      RecentLocalStorageItemsService
+    ) as jasmine.SpyObj<RecentLocalStorageItemsService>;
+    utilityService = TestBed.inject(UtilityService) as jasmine.SpyObj<UtilityService>;
+    orgSettingsService = TestBed.inject(OrgSettingsService) as jasmine.SpyObj<OrgSettingsService>;
+    orgUserSettingsService = TestBed.inject(OrgUserSettingsService) as jasmine.SpyObj<OrgUserSettingsService>;
+
+    orgUserSettingsService.get.and.returnValue(of(orgUserSettingsData));
+    projectService.getbyId.and.returnValue(
+      of({
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:108]: Move objects inside the data file

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+
+  describe('getProjects():', () => {
+    it('should get projects when current selection is not defined', (done) => {
+      orgSettingsService.get.and.returnValue(of(orgSettingsData));
+      authService.getEou.and.returnValue(Promise.resolve(apiEouRes));
+      orgUserSettingsService.get.and.returnValue(of(orgUserSettingsData));
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:172]: Some of these are already initialized in beforeEach, so are not needed here

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+      projectService.getByParamsUnformatted.and.returnValue(of(projects));
+      projectService.getbyId.and.returnValue(of(expectedProjects[0].value));
+
+      component.getProjects('projects').subscribe((res) => {
+        expect(res).toEqual(expectedProjects);
+        expect(orgSettingsService.get).toHaveBeenCalled();
+        expect(authService.getEou).toHaveBeenCalled();
+        expect(orgUserSettingsService.get).toHaveBeenCalled();
+        expect(projectService.getByParamsUnformatted).toHaveBeenCalledWith({
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:181]: `toHaveBeenCalled` -> `toHaveBeenCalledOnce`

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+        expect(projectService.getbyId).toHaveBeenCalledWith(3943);
+        done();
+      });
+    });
+
+    it('should get projects when current selection is defined', (done) => {
+      orgSettingsService.get.and.returnValue(of(orgSettingsData));
+      authService.getEou.and.returnValue(Promise.resolve(apiEouRes));
+      orgUserSettingsService.get.and.returnValue(of(orgUserSettingsData));
+      projectService.getByParamsUnformatted.and.returnValue(of(projects));
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:201]: Some initializations are already present in beforeEach

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+      projectService.getbyId.and.returnValue(of(expectedProjects[0].value));
+      component.currentSelection = [testProjectV2];
+      fixture.detectChanges();
+
+      component.getProjects('projects').subscribe((res) => {
+        expect(res).toEqual(expectedProjects2);
+        expect(orgSettingsService.get).toHaveBeenCalled();
+        expect(authService.getEou).toHaveBeenCalled();
+        expect(orgUserSettingsService.get).toHaveBeenCalled();
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:210]: Same comments as above

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+        });
+        expect(projectService.getbyId).toHaveBeenCalledWith(3943);
+        done();
+      });
+    });
+
+    it('should get projects when default value is null and no default projects are available', (done) => {
+      orgUserSettingsService.get.and.returnValue(of(orgUserSettingsData));
+      orgSettingsService.get.and.returnValue(of(orgSettingsDataWithoutAdvPro));
+      projectService.getbyId.and.returnValue(of(expectedProjects[0].value));
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:230]: same comments for this too

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+        expect(projectService.getbyId).toHaveBeenCalledWith(3943);
+        done();
+      });
+    });
+  });
+
+  it('clearValue(): should clear all values', () => {
+    component.value = 'value';
+
+    const dummyHtmlInputElement = document.createElement('input');
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:258]: Why create a new component? Use the searchBar we already have

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,443 @@
...
+    component.searchBarRef = {
+      nativeElement: dummyHtmlInputElement,
+    };
+    dummyHtmlInputElement.value = 'US';
+    dummyHtmlInputElement.dispatchEvent(new Event('keyup'));
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:263]: This should be an assetion

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+    fixture.detectChanges();
+
+    component.clearValue();
+    expect(component.value).toEqual('');
+    expect(dummyHtmlInputElement.value).toEqual('');
+  });
+
+  describe('getRecentlyUsedItems():', () => {
+    it('should return recently used project if already present', (done) => {
+      component.recentlyUsed = [
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:273]: What if recentlyUsed is an empty array? Check the behaviour is web app once and make changes in code if necessary

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+        {
+          label: 'label',
+          value: testProjectV2,
+        },
+      ];
+
+      fixture.detectChanges();
+
+      component.getRecentlyUsedItems().subscribe((res) => {
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:282]: Add assertion  recentLocalStorageItemsService.get should not have been called

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+            selected: false,
+          },
+        ]);
+        expect(recentLocalStorageItemsService.get).toHaveBeenCalledWith(component.cacheName);
+        done();
+      });
+    });
+  });
+
+  it('onDoneClick', async () => {
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:320]: async/await is not needed here

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts [hunk 12]
```diff
@@ -0,0 +1,443 @@
...
+    modalController.dismiss.and.returnValue(Promise.resolve(true));
+
+    await component.onDoneClick();
+    expect(modalController.dismiss).toHaveBeenCalledTimes(1);
+  });
+
+  describe('onElementSelect():', () => {
+    it('should dismiss the modal with selected option', async () => {
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:328]: same as above

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+        { value: 'value' },
+        'label'
+      );
+    });
+  });
+
+  it('ngAfterViewInit(): show filtered projects and recently used items', async () => {
+    spyOn(component, 'getRecentlyUsedItems').and.callThrough();
+    utilityService.searchArrayStream.and.returnValue(() => of([{ label: '', value: '' }]));
+    component.currentSelection = {
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:359]: Put large objects in data file

**Review by yksolanki9 — CHANGES_REQUESTED** [review]

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+    const inputElement = document.createElement('input');
+    component.searchBarRef = {
+      nativeElement: inputElement,
+    };
+
+    await component.ngAfterViewInit();
+    inputElement.value = 'projects';
+    inputElement.dispatchEvent(new Event('keyup'));
+
+    expect(component.getProjects).toHaveBeenCalledWith('projects');
```
[the remark this answers is not in the document]
Comment (jayfyle) [reply src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:390]: I am facing the date issue that keeps coming up once in a while, where the dates don't match and time conversion happens on karma but fails on github workflow.

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts (hunk 3 above)
Comment (jayfyle) [reply src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:181]: `ngAfterViewInit` is being called multiple times, as this function is called in the method, these methods are called more than once.

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts (hunk 5 above)
Comment (jayfyle) [reply src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:210]: Same, multiple calls to `ngAfterViewinit` is causing this.

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts (hunk 3 above)
Comment (yksolanki9) [reply src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:144]: This should not happen, it should be executed only once. Possibly some issue with the setup

File: src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts (hunk 7 above)
Comment (yksolanki9) [reply src/app/shared/components/fy-select-project/fy-select-modal/fy-select-project-modal.component.spec.ts:216]: This is not resolved

**Review by yksolanki9 — CHANGES_REQUESTED** [review]

**Review by yksolanki9 — APPROVED** [review]

**Pull request merged**